From f6343a3b7ca47ddcc81e2a39e2b16eb7f15cad1c Mon Sep 17 00:00:00 2001
From: Gerd Hoffmann <kraxel@redhat.com>
Date: Thu, 23 Jun 2011 12:41:50 -0300
Subject: [RHEL6 qemu-kvm PATCH 075/115] Bug #757654: UHCI fails to signal stall response patch

RH-Author: Gerd Hoffmann <kraxel@redhat.com>
Message-id: <1308832951-8995-75-git-send-email-kraxel@redhat.com>
Patchwork-id: 28395
O-Subject: [RHEL-6.2 kvm PATCH 074/115] Bug #757654: UHCI fails to signal stall response patch
Bugzilla: 561414 632299 645351 711354
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>
RH-Acked-by: Kevin Wolf <kwolf@redhat.com>

From: Jan Vesely <jano.vesely@gmail.com>

UHCI host controller status register indicates error and
an interrupt is triggered on BABBLE and STALL errors.

Signed-off-by: Jan Vesely <jano.vesely@gmail.com>
Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>
(cherry picked from commit 8656954aedbd9995e68e998df734a849f8e63ade)
---
 hw/usb-uhci.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

Signed-off-by: Eduardo Habkost <ehabkost@redhat.com>
---
 hw/usb-uhci.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/hw/usb-uhci.c b/hw/usb-uhci.c
index 429ac2c..1505fac 100644
--- a/hw/usb-uhci.c
+++ b/hw/usb-uhci.c
@@ -699,11 +699,15 @@ out:
     case USB_RET_STALL:
         td->ctrl |= TD_CTRL_STALL;
         td->ctrl &= ~TD_CTRL_ACTIVE;
+        s->status |= UHCI_STS_USBERR;
+        uhci_update_irq(s);
         return 1;
 
     case USB_RET_BABBLE:
         td->ctrl |= TD_CTRL_BABBLE | TD_CTRL_STALL;
         td->ctrl &= ~TD_CTRL_ACTIVE;
+        s->status |= UHCI_STS_USBERR;
+        uhci_update_irq(s);
         /* frame interrupted */
         return -1;
 
-- 
1.7.3.2

